In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.ensemble import BaggingClassifier
from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import AdaBoostClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn import metrics

In [2]:
data = pd.read_csv('/content/Company_Data.csv')

In [3]:
data.head()

Sales  CompPrice  Income  Advertising  ...  Age  Education Urban   US
0   9.50        138      73           11  ...   42         17   Yes  Yes
1  11.22        111      48           16  ...   65         10   Yes  Yes
2  10.06        113      35           10  ...   59         12   Yes  Yes
3   7.40        117     100            4  ...   55         14   Yes  Yes
4   4.15        141      64            3  ...   38         13   Yes   No

[5 rows x 11 columns]

In [4]:
data.corr()

Sales  CompPrice    Income  ...     Price       Age  Education
Sales        1.000000   0.064079  0.151951  ... -0.444951 -0.231815  -0.051955
CompPrice    0.064079   1.000000 -0.080653  ...  0.584848 -0.100239   0.025197
Income       0.151951  -0.080653  1.000000  ... -0.056698 -0.004670  -0.056855
Advertising  0.269507  -0.024199  0.058995  ...  0.044537 -0.004557  -0.033594
Population   0.050471  -0.094707 -0.007877  ... -0.012144 -0.042663  -0.106378
Price       -0.444951   0.584848 -0.056698  ...  1.000000 -0.102177   0.011747
Age         -0.231815  -0.100239 -0.004670  ... -0.102177  1.000000   0.006488
Education   -0.051955   0.025197 -0.056855  ...  0.011747  0.006488   1.000000

[8 rows x 8 columns]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Sales        400 non-null    float64
 1   CompPrice    400 non-null    int64  
 2   Income       400 non-null    int64  
 3   Advertising  400 non-null    int64  
 4   Population   400 non-null    int64  
 5   Price        400 non-null    int64  
 6   ShelveLoc    400 non-null    object 
 7   Age          400 non-null    int64  
 8   Education    400 non-null    int64  
 9   Urban        400 non-null    object 
 10  US           400 non-null    object 
dtypes: float64(1), int64(7), object(3)
memory usage: 34.5+ KB


In [6]:
data.describe()

Sales   CompPrice      Income  ...       Price         Age   Education
count  400.000000  400.000000  400.000000  ...  400.000000  400.000000  400.000000
mean     7.496325  124.975000   68.657500  ...  115.795000   53.322500   13.900000
std      2.824115   15.334512   27.986037  ...   23.676664   16.200297    2.620528
min      0.000000   77.000000   21.000000  ...   24.000000   25.000000   10.000000
25%      5.390000  115.000000   42.750000  ...  100.000000   39.750000   12.000000
50%      7.490000  125.000000   69.000000  ...  117.000000   54.500000   14.000000
75%      9.320000  135.000000   91.000000  ...  131.000000   66.000000   16.000000
max     16.270000  175.000000  120.000000  ...  191.000000   80.000000   18.000000

[8 rows x 8 columns]

In [7]:
data['Status'] = data['Sales'].apply(lambda Sales: 'Low' if Sales <= 7.5 else 'High')

In [8]:
data.head()

Sales  CompPrice  Income  Advertising  ...  Education  Urban   US  Status
0   9.50        138      73           11  ...         17    Yes  Yes    High
1  11.22        111      48           16  ...         10    Yes  Yes    High
2  10.06        113      35           10  ...         12    Yes  Yes    High
3   7.40        117     100            4  ...         14    Yes  Yes     Low
4   4.15        141      64            3  ...         13    Yes   No     Low

[5 rows x 12 columns]

In [9]:
data.drop(['Sales'], axis=1, inplace=True)

In [10]:
data.head()

CompPrice  Income  Advertising  Population  ...  Education Urban   US  Status
0        138      73           11         276  ...         17   Yes  Yes    High
1        111      48           16         260  ...         10   Yes  Yes    High
2        113      35           10         269  ...         12   Yes  Yes    High
3        117     100            4         466  ...         14   Yes  Yes     Low
4        141      64            3         340  ...         13   Yes   No     Low

[5 rows x 11 columns]

In [11]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 400 entries, 0 to 399
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   CompPrice    400 non-null    int64 
 1   Income       400 non-null    int64 
 2   Advertising  400 non-null    int64 
 3   Population   400 non-null    int64 
 4   Price        400 non-null    int64 
 5   ShelveLoc    400 non-null    object
 6   Age          400 non-null    int64 
 7   Education    400 non-null    int64 
 8   Urban        400 non-null    object
 9   US           400 non-null    object
 10  Status       400 non-null    object
dtypes: int64(7), object(4)
memory usage: 34.5+ KB


In [12]:
from sklearn.preprocessing import LabelEncoder

In [15]:
encoder = LabelEncoder()
for i in data.columns:
    print(i)
    if data[i].dtypes == object:
        data[i]= encoder.fit_transform(data[i])
    else:
        pass

CompPrice
Income
Advertising
Population
Price
ShelveLoc
Age
Education
Urban
US
Status


In [16]:
data.head()

CompPrice  Income  Advertising  Population  ...  Education  Urban  US  Status
0        138      73           11         276  ...         17      1   1       0
1        111      48           16         260  ...         10      1   1       0
2        113      35           10         269  ...         12      1   1       0
3        117     100            4         466  ...         14      1   1       1
4        141      64            3         340  ...         13      1   0       1

[5 rows x 11 columns]

High is 0, Low is 1

In [27]:
y = data.iloc[:, -1]
y.value_counts()

1    202
0    198
Name: Status, dtype: int64

In [29]:
array = data.values
X = array[:, 0:10]
Y = array[:, -1]
seed=7
kfold = KFold(n_splits=8, random_state=seed)

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:296: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning


In [22]:
X

array([[138,  73,  11, ...,  17,   1,   1],
       [111,  48,  16, ...,  10,   1,   1],
       [113,  35,  10, ...,  12,   1,   1],
       ...,
       [162,  26,  12, ...,  18,   1,   1],
       [100,  79,   7, ...,  12,   1,   1],
       [134,  37,   0, ...,  16,   1,   1]])

In [28]:
X_train, X_test, Y_train, Y_test = train_test_split(X,Y, test_size= 0.3)

#Bagging Method

In [44]:
cart = DecisionTreeClassifier()
num_trees= 100
model = BaggingClassifier(base_estimator=cart, n_estimators= num_trees, random_state=seed)
model.fit(X_train, Y_train)
results = cross_val_score(model, X, Y, cv= kfold)
print(results.mean())

0.805


In [45]:
#Predicting on the test data
preds = model.predict(X_test)
pd.Series(preds).value_counts()

0    66
1    54
dtype: int64

In [46]:
pd.crosstab(Y_test, preds)

col_0    0   1
Status        
0       53  13
1       13  41

In [47]:
print('Accuracy: ', np.mean(preds==Y_test))

Accuracy:  0.7833333333333333


Bagging method accuracy on the test data is 78.33%

#Random Forest

In [67]:
num_trees = 100
max_features = 3
model1 = RandomForestClassifier(n_estimators=num_trees, max_features=max_features)
model1.fit(X_train, Y_train)
results1 = cross_val_score(model1, X, Y, cv=kfold)
print(results1.mean())

0.8125


In [68]:
#Predicting on the rest data
preds1 = model1.predict(X_test)
pd.Series(preds1).value_counts()

1    61
0    59
dtype: int64

In [69]:
pd.crosstab(Y_test, preds1)

col_0    0   1
Status        
0       48  18
1       11  43

In [70]:
print("Accuracy: ", np.mean(preds1==Y_test))

Accuracy:  0.7583333333333333


Random Forest accuracy on test data is 75%

#Adaboost

In [58]:

num_trees = 20
kfold= KFold(n_splits=8, random_state=seed)
model2= AdaBoostClassifier(n_estimators=num_trees, random_state=seed)
model2.fit(X_train, Y_train)
results2 = cross_val_score(model3, X, Y, cv=kfold)
print(results2.mean())

/usr/local/lib/python3.7/dist-packages/sklearn/model_selection/_split.py:296: FutureWarning: Setting a random_state has no effect since shuffle is False. This will raise an error in 0.24. You should leave random_state to its default (None), or set shuffle=True.
  FutureWarning


0.8225


In [59]:
#Predicting on the rest data
preds2 = model2.predict(X_test)
pd.Series(preds2).value_counts()

0    62
1    58
dtype: int64

In [71]:
pd.crosstab(Y_test, preds2)

col_0    0   1
Status        
0       56  10
1        6  48

In [72]:
print("Accuracy: ", np.mean(preds2==Y_test))

Accuracy:  0.8666666666666667


Adaboost accuracy on test data is 86.66%
####Adaboost is the best model to analyse the Sale of the company